pkg/common/cns-lib/volume/manager.go

```diff
@@ -315,23 +315,59 @@ func (m *defaultManager) createVolumeWithImprovedIdempotency(ctx context.Context
 		}
 	}
 
-	// Get the taskInfo.
-	taskInfo, err := cns.GetTaskInfo(ctx, task)
-	if err != nil || taskInfo == nil {
-		log.Errorf("failed to get taskInfo for CreateVolume task with err: %v", err)
-		if err != nil {
-			faultType = ExtractFaultTypeFromErr(ctx, err)
-		} else {
-			faultType = csifault.CSITaskInfoEmptyFault
+	taskInfo, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		// WaitForResult can fail for many reasons, including:
+		// - Task was completed and deleted from VCDB.
+		// - CNS restarted and marked "InProgress" tasks as "Failed".
+		// In all cases, query CNS to determine if the volume was created successfully.
+		log.Debugf("Failed to get task result for task %s with error %v. Querying CNS "+
+			"to determine if the volume %s was successfully created.",
+			task.Reference().Value, err, volNameFromInputSpec)
+		queryFilter := cnstypes.CnsQueryFilter{
+			Names:               []string{volNameFromInputSpec},
+			ContainerClusterIds: []string{spec.Metadata.ContainerClusterArray[0].ClusterId},
+		}
+		querySelection := cnstypes.CnsQuerySelection{
+			Names: []string{
+				string(cnstypes.QuerySelectionNameTypeVolumeName),
+			},
 		}
-		return nil, faultType, err
+		queryResult, queryAllVolumeErr := m.QueryAllVolume(ctx, queryFilter, querySelection)
```

Comment:
This is redundant if the CNS Create volume failed with some error.

Reply:
Yes, but how do i differentiate here from errors due to the network or FCD layer etc, where the creation may have succeeded?

Reply:
As per offline discussion, only querying CNS if the task is deleted from VCDB.